pkg/chunk/storage/storage_client_test.go

```diff
@@ -88,6 +106,135 @@ func testStorageClientChunks(t *testing.T, client chunk.StorageClient) {
 	}
 }
 
+func testStorageClientIndex(t *testing.T, client chunk.StorageClient, tableClient chunk.TableClient) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	type writeBatch []struct {
+		tableName, hashValue string
+		rangeValue           []byte
+		value                []byte
+	}
+
+	cases := []struct {
+		batch writeBatch
+
+		query []chunk.IndexQuery
+	}{
+		{
+			batch: writeBatch{
+				{
+					tableName:  "t1",
+					hashValue:  "hash1",
+					rangeValue: []byte("range1"),
+					value:      []byte("1"),
+				},
+				{
+					tableName:  "t1",
+					hashValue:  "hash1",
+					rangeValue: []byte("range2"),
+					value:      []byte("2"),
+				},
+				{
+					tableName:  "t1",
+					hashValue:  "hash1",
+					rangeValue: []byte("range3"),
+					value:      []byte("3"),
+				},
+				{
+					tableName:  "t1",
+					hashValue:  "hash1",
+					rangeValue: []byte("bleepboop"),
+					value:      []byte("4"),
+				},
+				{
+					tableName:  "t1",
+					hashValue:  "hash2",
+					rangeValue: []byte("bleepboop"),
+					value:      []byte("5"),
+				},
+			},
+
+			query: []chunk.IndexQuery{
+				{
+					TableName:        "t1",
+					HashValue:        "hash1",
+					RangeValuePrefix: []byte("range"),
+				},
+				{
+					TableName:       "t1",
+					HashValue:       "hash1",
+					RangeValueStart: []byte("range2"),
+				},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		mockSC := chunk.NewMockStorage()
+		require.NoError(t, mockSC.CreateTable(ctx, chunk.TableDesc{
+			Name: c.batch[0].tableName,
+		}))
+		require.NoError(t, tableClient.CreateTable(ctx, chunk.TableDesc{
+			Name:             c.batch[0].tableName,
+			ProvisionedRead:  100000,
+			ProvisionedWrite: 100000,
+		}))
```

Comment:
Not a huge fan of using the mock store to get the expected result; I think I'd prefer having the test case spell out explicitly what we expect to return.

Reply:
Hmm, it actually increases the test verbosity by a fair share. I'd rather add test-cases to increase the confidence in the in-mem version, but it's your call.